gcoap: multi-transport support

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
https://github.com/RIOT-OS/RIOT/pull/15549 introduced DTLS support for CoAP, but did not make it possible to select the transport on request. Since switching between CoAP and CoAPS (CoAP-over-DTLS) as client is a valid use case (one might want to e.g. talk to one server over CoAP and to another over CoAPS), this change makes that possible.

This is only a draft for now.

TODOs:

- [ ] Testing.
- [x] Also make listeners configurable at run-time.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
Untested for now.
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Follow-up on https://github.com/RIOT-OS/RIOT/pull/15549
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

miri64 commented:
That was supposed to be a draft PR... Oh well... clicked to quickly.

miri64 commented:
Sending messages using both the `gcoap` and `gcoap_dtls` example seems to work now. I can also observe the respective `/cli/stats` resource using `aiocoap`.

miri64 commented:
I added support for listeners now, but I am not 100% happy yet, as `/.well-known/core` still advertises all resources, regardless if they are available with the used transport or not. To fix this, the only thing I can think of at the moment would require to expose the transport type to the request handler, but that I would prefer not to do, as this would increase the number of parameters (meaning ARM and co would start using the stack instead of the general purpose registers for them) and currently this is the only use-case I can think of where this information could be useful (as usually one knows which resource they registered for which transport).

miri64 commented:
I know there is [`draft-amsuess-core-transport-indication`](https://www.ietf.org/archive/id/draft-amsuess-core-transport-indication-01.html), so maybe with that, this becomes a non-issue. But maybe @chrysn can give some insight here.

miri64 commented:
Heavily edited https://github.com/RIOT-OS/RIOT/pull/16688#issuecomment-888374420 to provide some better context.

miri64 commented:
Input on the proposed design is welcome.

chrysn commented:
Sorry, I'll get to it at earliest in 10 days :-(

benpicco commented on examples/gcoap/gcoap_cli.c:
```diff
@@ -78,6 +78,7 @@ static const char *_link_params[] = {
 static gcoap_listener_t _listener = {
     &_resources[0],
     ARRAY_SIZE(_resources),
+    GCOAP_SOCKET_TYPE_UNSPEC,
```
```suggestion
    GCOAP_SOCKET_TYPE_UNDEF,
```

benpicco reviewed (commented):
First test with a `gcoap_dtls` server and both a `gcoap_dtls` and `gcoap` client is successful.
Both can access resources on the server via UDP and DTLS.

With this small change

```patch
--- a/examples/gcoap/gcoap_cli.c
+++ b/examples/gcoap/gcoap_cli.c
@@ -67,6 +67,9 @@ static ssize_t _riot_board_handler(coap_pkt_t* pdu, uint8_t *buf, size_t len, vo
 /* CoAP resources. Must be sorted by path (ASCII order). */
 static const coap_resource_t _resources[] = {
     { "/cli/stats", COAP_GET | COAP_PUT, _stats_handler, NULL },
+};
+
+static const coap_resource_t _resources_secure[] = {
     { "/riot/board", COAP_GET, _riot_board_handler, NULL },
 };
 
@@ -78,7 +81,16 @@ static const char *_link_params[] = {
 static gcoap_listener_t _listener = {
     &_resources[0],
     ARRAY_SIZE(_resources),
-    GCOAP_SOCKET_TYPE_UNSPEC,
+    GCOAP_SOCKET_TYPE_UNDEF,
+    _encode_link,
+    NULL,
+    NULL
+};
+
+static gcoap_listener_t _listener_secure = {
+    &_resources_secure[0],
+    ARRAY_SIZE(_resources_secure),
+    GCOAP_SOCKET_TYPE_DTLS,
     _encode_link,
     NULL,
     NULL
@@ -491,4 +503,5 @@ void gcoap_cli_init(void)
 #endif
 
     gcoap_register_listener(&_listener);
+    gcoap_register_listener(&_listener_secure);
 }
```

The `/riot/board` resource is only available using DTLS while `/cli/stats` remains accessible using both UDP and DTLS.

It's unfortunately not possible for the `gcoap_dtls` example to make unencrypted requests, but that's a limitation of the CLI command. Would probably be better if the `coap` command accepted an URI instead of `<addr>[%iface] <port> <path>` but this is for another PR.

miri64 commented:
> Sorry, I'll get to it at earliest in 10 days :-(

@chrysn ping?

miri64 commented:
Ok, I now added the transport filter for `.well-known/core`, but I think it is a bit hacky: For the `_well_known_core_handler` a dynamic context object is now added (the `gcoap_socket_t` of the current connection), while the context behavior for all other resources stays static and is taken from the `coap_resource_t` struct's `context` pointer, so as it was before.

benpicco commented:
> I now added the transport filter for `.well-known/core`

Why does that need special handling?

miri64 commented:
> > I now added the transport filter for `.well-known/core`
> 
> Why does that need special handling?

See https://github.com/RIOT-OS/RIOT/pull/16688#issuecomment-888374420

benpicco commented:
I wouldn't find it too bad if all handlers could know whether they've been accessed via DTLS or plain UDP.
That way the same handler could return different results based on the transport.

benpicco commented:
Another option would be to just duplicate the endpoint

```C
/* Internal variables */
const coap_resource_t _default_resources[] = {
    { "/.well-known/core", COAP_GET, _well_known_core_handler, GCOAP_SOCKET_TYPE_UDP },
};

const coap_resource_t _default_secure_resources[] = {
    { "/.well-known/core", COAP_GET, _well_known_core_handler, GCOAP_SOCKET_TYPE_DTLS },
};

static gcoap_listener_t _default_listener = {
    &_default_resources[0],
    ARRAY_SIZE(_default_resources),
    GCOAP_SOCKET_TYPE_UDP,
    NULL,
    NULL,
    _request_matcher_default
};

static gcoap_listener_t _default_secure_listener = {
    &_default_resources[0],
    ARRAY_SIZE(_default_secure_resources),
    GCOAP_SOCKET_TYPE_DTLS,
    NULL,
    NULL,
    _request_matcher_default
};
```

benpicco commented:
Do you think we can get this into the 2021.10 release?

miri64 commented:
> Do you think we can get this into the 2021.10 release?

No time to get it out of WIP ATM + I to me it make sense to wait at least until #16855 got somewhat started so I can put the new struct member of `coap_pkt_t` directly into `gcoap_pkt_t`, so we don't have to endure a deprecation round for that as well. So I would tend to say bump (even though it is still some time for hard feature freeze).

benpicco commented:
is there anything that can be done to help with this PR?

miri64 commented:
> is there anything that can be done to help with this PR?

As stated in https://github.com/RIOT-OS/RIOT/pull/16688#issuecomment-920790414, we should work on #16855. I started some of that work in #17168.

miri64 commented:
> is there anything that can be done to help with this PR?

Other than that, some testing application would be great!

benpicco commented:
Want to give this a rebase? Feel free to squash while you're at it.

miri64 commented:
Rebased and it at least still compiles. However, with the recent rework in #17471, the `gcoap_dtls` now does not seem to work in current master (the client sends out unencrypted CoAP messages), and as such I was unable to test if my rebase worked. I was very careful, so I don't think I broke anything. IMHO, once we can test it (i.e. the error in allegedly introduced in #17471 is fixed), this is ready to merge: Sure we can wait until we reworked and split nanocoap and gcoap, but lets not be perfect the enemy of good enough ;-).

miri64 commented:
Ok, I think I was just tired yesterday. I tested it again today, and now it works

```
RIOT native interrupts/signals initialized.
RIOT native board initialized.
RIOT native hardware initialization complete.

NETOPT_RX_END_IRQ not implemented by driver
main(): This is RIOT! (Version: 2022.04-devel-577-g287bfd-gcoap/enh/multitransport)
gcoap example app
All up, running the shell now
> ifconfig
ifconfig
Iface  7  HWaddr: DA:27:1D:A8:64:24 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::d827:1dff:fea8:6424  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffa8:6424
          
> coap info
coap info
CoAP server is listening on port 5684
Connection secured with DTLS
Free DTLS session slots: 1/1
 CLI requests sent: 0
CoAP open requests: 0
Configured Proxy: None
```


```
RIOT native interrupts/signals initialized.
RIOT native board initialized.
RIOT native hardware initialization complete.

NETOPT_RX_END_IRQ not implemented by driver
main(): This is RIOT! (Version: 2022.04-devel-577-g287bfd-gcoap/enh/multitransport)
gcoap example app
All up, running the shell now
> coap get fe80::d827:1dff:fea8:6424 5684 /.well-known/core
coap get fe80::d827:1dff:fea8:6424 5684 /.well-known/core
gcoap_cli: sending msg ID 61972, 23 bytes
> gcoap: response Success, code 2.05, 46 bytes
</cli/stats>;ct=0;rt="count";obs,</riot/board>
```
![image](https://user-images.githubusercontent.com/675644/156375098-52c90a98-f152-486b-a783-262a3958caf6.png)

miri64 commented:
:tada:
